File: agent/limits.py
"""
Rate-limit windows — the real ones, read from Claude Code's own cache.

Claude Code asks Anthropic what fraction of your plan you have used and when
each window rolls over, and writes the answer to `~/.claude.json` under
`cachedUsageUtilization`. It is the same body its own Account & Usage panel
renders. So the percentages here are not derived, not estimated, and not
inferred from local activity — they are the server's numbers, forwarded.

Note the path: `~/.claude.json`, a file, not something inside `~/.claude/`.

Three things shape this module, and all three are honesty problems rather
than engineering ones:

  · **It is a cache, so it has an age.** Nothing here can refresh it. It is
    written as a side effect of a Claude Code session hitting the usage
    endpoint, and the CLI discards it after an hour. So the age is reported
    with the reading, always, and it is measured from `fetchedAtMs` — never
    from the file's mtime, which moves when unrelated settings are written
    and would make a stale number look fresh.

  · **A reset instant does not go stale.** `resets_at` is an absolute time.
    When the percentage is too old to trust, the countdown beside it is still
    exactly right. The two degrade separately and the payload keeps them
    separable.

  · **The file holds far more than this.** Real name, email address,
    organisation name and role, billing tier, MCP configuration, and a
    `projects` map with per-project cost and token history. This reads
    exactly one key out of it. The omissions are deliberate, not oversights:

      - `utilization.spend` and `utilization.extra_usage` are actual billed
        dollars on accounts with extra usage enabled. The board's one estimate
        is labelled as an estimate everywhere it appears, and a real bill
        rendered next to it in the same style would destroy that distinction.
        If it is ever shown it needs its own panel and its own words.
      - `projects` carries a second, richer usage history — lastCost,
        lastModelUsage, per-project token totals. The board reads usage from
        the transcripts instead, which is current rather than last-session.
      - `oauthAccount` is never touched. The file is mode 0600 and that is
        the OS agreeing it is private.

    It never writes: that file is Claude Code's live configuration, we do not
    own it, and a clobbering write would take the user's MCP servers with it.

Keys are read BY NAME, never enumerated. `utilization` carries a dozen
unshipped internal buckets under rotating codenames; a panel built by
iterating that dict would one day grow a meter labelled "amber ladder".
"""
from __future__ import annotations

import hashlib
import json
import os
import secrets
import time
from datetime import datetime, timezone
from pathlib import Path

# The CLI's own reader discards this cache once it is an hour old. Matching
# that number means the board calls stale exactly what Claude Code calls
# stale, rather than inventing a second opinion.
STALE_AFTER = 3600

LABELS = {
    "session": "Session (5h)",
    "weekly_all": "Weekly (7 day)",
}


def _salt() -> str:
    """A per-install salt for the account hash.

    Without it, a truncated hash of the account uuid is the same string on
    every machine that account touches — a stable cross-host identifier
    travelling in a payload that crosses a network. The point of the hash is
    to answer "is this still the same account as last reading" on ONE board,
    and a salt kept beside the agent's other state does that without
    producing a value that correlates anything anywhere else.
    """
    path = Path(os.environ.get("TOKENHUD_STATE", Path.home() / ".tokenhud")).expanduser() / "salt"
    try:
        return path.read_text().strip()
    except Exception:
        pass
    value = secrets.token_hex(16)
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(value)
        os.chmod(path, 0o600)
    except Exception:
        pass
    return value


def _candidates() -> list[Path]:
    out = []
    env = os.environ.get("TOKENHUD_CLAUDE_JSON")
    if env:
        out.append(Path(env).expanduser())
    cfg = os.environ.get("CLAUDE_CONFIG_DIR")
    if cfg:
        out.append(Path(cfg).expanduser() / ".claude.json")
    out.append(Path.home() / ".claude.json")
    return out


def _load() -> dict | None:
    """Read the config, tolerating the moment it is being rewritten.

    Claude Code rewrites this file while running. A read landing mid-write
    gets a JSONDecodeError, which is a transient fact about timing and not a
    fact about the machine — so it is retried once before being reported.
    """
    for path in _candidates():
        if not path.is_file():
            continue
        for attempt in (0, 1):
            try:
                with path.open("rb") as fh:
                    return {"path": str(path), "data": json.loads(fh.read())}
            except json.JSONDecodeError:
                if attempt == 0:
                    time.sleep(0.1)
                    continue
                return {"path": str(path), "error": "unreadable"}
            except Exception:
                return {"path": str(path), "error": "unreadable"}
    return None


def _label(row: dict) -> str:
    kind = row.get("kind")
    if kind in LABELS:
        return LABELS[kind]
    if kind == "weekly_scoped":
        # Only the display name the server itself supplied. The neighbouring
        # keys in this file are internal bucket codenames that rotate, and
        # nothing on disk maps them to a model — inventing that mapping would
        # put a made-up model name on the board.
        name = ((row.get("scope") or {}).get("model") or {}).get("display_name")
        return f"Weekly · {name}" if name else "Weekly · scoped"
    return str(kind or "window").replace("_", " ").title()


def _rows(util: dict) -> list:
    rows = []
    for row in (util.get("limits") or []):
        if not isinstance(row, dict):
            continue
        pct = row.get("percent")
        rows.append({
            "kind": row.get("kind"),
            "group": row.get("group"),
            "label": _label(row),
            "percent": int(pct) if isinstance(pct, (int, float)) else None,
            "severity": row.get("severity") or "normal",
            "resetsAt": row.get("resets_at"),
            "active": bool(row.get("is_active")),
        })
    if rows:
        return rows

    # Older CLIs wrote only these two. Same numbers, less structure.
    for key, kind in (("five_hour", "session"), ("seven_day", "weekly_all")):
        w = util.get(key)
        if not isinstance(w, dict):
            continue
        pct = w.get("utilization")
        rows.append({
            "kind": kind, "group": kind.split("_")[0],
            "label": LABELS.get(kind, kind),
            "percent": int(pct) if isinstance(pct, (int, float)) else None,
            "severity": "normal",
            "resetsAt": w.get("resets_at"),
            "active": kind == "session",
        })
    return rows


def collect_limits() -> dict:
    """The plan's usage windows as Anthropic last reported them."""
    found = _load()
    if not found:
        return {"available": False, "reason": "absent", "staleAfterSeconds": STALE_AFTER}
    if found.get("error"):
        return {"available": False, "reason": found["error"], "staleAfterSeconds": STALE_AFTER}

    cache = (found["data"] or {}).get("cachedUsageUtilization")
    if not isinstance(cache, dict):
        return {"available": False, "reason": "absent", "staleAfterSeconds": STALE_AFTER}

    util = cache.get("utilization")
    if not isinstance(util, dict):
        return {"available": False, "reason": "absent", "staleAfterSeconds": STALE_AFTER}

    fetched_ms = cache.get("fetchedAtMs")
    fetched_at, age = None, None
    if isinstance(fetched_ms, (int, float)):
        fetched_at = datetime.fromtimestamp(fetched_ms / 1000, tz=timezone.utc).isoformat()
        age = int(time.time() - fetched_ms / 1000)

    acct = cache.get("accountUuid")
    return {
        "available": True,
        # Named so the board can say where a number came from without the
        # reader having to take it on trust.
        "source": "~/.claude.json:cachedUsageUtilization",
        "fetchedAt": fetched_at,
        "ageSeconds": age,
        "staleAfterSeconds": STALE_AFTER,
        # Answers "is this the same account as last reading" on this board,
        # and nothing else anywhere else. Salted per install so it is not a
        # cross-host identifier for the account it stands in for.
        "accountHash": (hashlib.sha256((_salt() + str(acct)).encode()).hexdigest()[:12]
                        if acct else None),
        "windows": _rows(util),
        # Deliberately not computed here: a snapshot can sit in the store for
        # minutes before a browser reads it, and a countdown baked in at
        # collection time would be wrong by exactly that long. The browser
        # holds the clock; this ships the instant.
    }


if __name__ == "__main__":
    print(json.dumps(collect_limits(), indent=2))

File: agent/src/limits.rs
//! Rate-limit windows - the real ones, read from Claude Code's own cache.
//!
//! Three things shape this module, and all three are honesty problems rather
//! than engineering ones:
//!
//!   · **It is a cache, so it has an age.** Measured from `fetchedAtMs`, never
//!     from the file's mtime, which moves when unrelated settings are written
//!     and would make a stale number look fresh.
//!   · **A reset instant does not go stale.** `resets_at` is absolute; the two
//!     degrade separately and the payload keeps them separable.
//!   · **The file holds far more than this, and the omissions are deliberate.**
//!     `utilization.spend`, `projects`, `oauthAccount` are never read. Keys are
//!     read BY NAME, never enumerated - `utilization` carries unshipped internal
//!     buckets under rotating codenames, and a panel built by iterating that
//!     dict would one day grow a meter labelled "amber ladder".
//!
//! It never writes to that file. It is Claude Code's live configuration, we do
//! not own it, and a clobbering write would take the user's MCP servers with it.

use crate::transcripts::{expand_tilde, home, state_dir};
use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};
use std::fs;
use std::path::PathBuf;

/// The CLI's own reader discards this cache once it is an hour old. Matching
/// that means the board calls stale exactly what Claude Code calls stale.
pub const STALE_AFTER: i64 = 3600;

fn label_for(kind: &str) -> Option<&'static str> {
    match kind {
        "session" => Some("Session (5h)"),
        "weekly_all" => Some("Weekly (7 day)"),
        _ => None,
    }
}

/// A per-install salt for the account hash.
///
/// Without it, a truncated hash of the account uuid is the same string on every
/// machine that account touches - a stable cross-host identifier travelling in
/// a payload that crosses a network. The point of the hash is to answer "is
/// this still the same account as last reading" on ONE board.
fn salt() -> String {
    let path = state_dir().join("salt");
    if let Ok(s) = fs::read_to_string(&path) {
        let s = s.trim().to_string();
        if !s.is_empty() {
            return s;
        }
    }
    let value = random_hex(16);
    if let Some(dir) = path.parent() {
        let _ = fs::create_dir_all(dir);
    }
    if fs::write(&path, &value).is_ok() {
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let _ = fs::set_permissions(&path, fs::Permissions::from_mode(0o600));
        }
    }
    value
}

/// 16 bytes of OS randomness, hex. `getrandom` rather than a PRNG crate: the
/// salt is written once and a dependency for one call is not worth carrying.
/// Pub because the install id (main.rs) is minted the same way for the same
/// reasons.
pub fn random_hex(n: usize) -> String {
    let mut buf = vec![0u8; n];
    let ok = unsafe { libc::getentropy(buf.as_mut_ptr() as *mut libc::c_void, buf.len()) } == 0;
    if !ok {
        // Never silently produce a constant salt: fall back to something that
        // still differs per install rather than to zeros.
        let t = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or(0);
        let pid = std::process::id() as u128;
        let mut h = Sha256::new();
        h.update(t.to_le_bytes());
        h.update(pid.to_le_bytes());
        buf.copy_from_slice(&h.finalize()[..n]);
    }
    buf.iter().map(|b| format!("{b:02x}")).collect()
}

fn candidates() -> Vec<PathBuf> {
    let mut out = Vec::new();
    if let Ok(v) = std::env::var("TOKENHUD_CLAUDE_JSON") {
        if !v.is_empty() {
            out.push(expand_tilde(&v));
        }
    }
    if let Ok(v) = std::env::var("CLAUDE_CONFIG_DIR") {
        if !v.is_empty() {
            out.push(expand_tilde(&v).join(".claude.json"));
        }
    }
    out.push(home().join(".claude.json"));
    out
}

enum Found {
    Data(Value),
    Unreadable,
    Absent,
}

/// Read the config, tolerating the moment it is being rewritten.
///
/// Claude Code rewrites this file while running. A read landing mid-write gets
/// a parse error, which is a transient fact about timing and not a fact about
/// the machine - so it is retried once before being reported.
fn load() -> Found {
    for path in candidates() {
        if !path.is_file() {
            continue;
        }
        for attempt in 0..2 {
            match fs::read(&path) {
                Ok(bytes) => match serde_json::from_slice::<Value>(&bytes) {
                    Ok(v) => return Found::Data(v),
                    Err(_) if attempt == 0 => {
                        std::thread::sleep(std::time::Duration::from_millis(100));
                        continue;
                    }
                    Err(_) => return Found::Unreadable,
                },
                Err(_) => return Found::Unreadable,
            }
        }
    }
    Found::Absent
}

fn label(row: &Map<String, Value>) -> String {
    let kind = row.get("kind").and_then(|v| v.as_str()).unwrap_or("");
    if let Some(l) = label_for(kind) {
        return l.to_string();
    }
    if kind == "weekly_scoped" {
        // Only the display name the server itself supplied. The neighbouring
        // keys are internal bucket codenames that rotate, and nothing on disk
        // maps them to a model - inventing that mapping would put a made-up
        // model name on the board.
        let name = row
            .get("scope")
            .and_then(|s| s.get("model"))
            .and_then(|m| m.get("display_name"))
            .and_then(|v| v.as_str());
        return match name {
            Some(n) => format!("Weekly · {n}"),
            None => "Weekly · scoped".to_string(),
        };
    }
    if kind.is_empty() {
        return title_case("window");
    }
    title_case(&kind.replace('_', " "))
}

fn title_case(s: &str) -> String {
    s.split(' ')
        .map(|w| {
            let mut c = w.chars();
            match c.next() {
                Some(f) => f.to_uppercase().collect::<String>() + &c.as_str().to_lowercase(),
                None => String::new(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

fn rows(util: &Value) -> Vec<Value> {
    let mut out = Vec::new();
    if let Some(list) = util.get("limits").and_then(|v| v.as_array()) {
        for row in list {
            let row = match row.as_object() {
                Some(o) => o,
                None => continue,
            };
            let pct = row
                .get("percent")
                .and_then(|v| v.as_f64())
                .map(|f| f as i64);
            out.push(json!({
                "kind": row.get("kind").cloned().unwrap_or(Value::Null),
                "group": row.get("group").cloned().unwrap_or(Value::Null),
                "label": label(row),
                "percent": pct,
                "severity": row.get("severity").and_then(|v| v.as_str()).unwrap_or("normal"),
                "resetsAt": row.get("resets_at").cloned().unwrap_or(Value::Null),
                "active": row.get("is_active").and_then(|v| v.as_bool()).unwrap_or(false),
            }));
        }
    }
    if !out.is_empty() {
        return out;
    }

    // Older CLIs wrote only these two. Same numbers, less structure.
    for (key, kind) in [("five_hour", "session"), ("seven_day", "weekly_all")] {
        let w = match util.get(key) {
            Some(w) if w.is_object() => w,
            _ => continue,
        };
        let pct = w
            .get("utilization")
            .and_then(|v| v.as_f64())
            .map(|f| f as i64);
        out.push(json!({
            "kind": kind,
            "group": kind.split('_').next().unwrap_or(kind),
            "label": label_for(kind).unwrap_or(kind),
            "percent": pct,
            "severity": "normal",
            "resetsAt": w.get("resets_at").cloned().unwrap_or(Value::Null),
            "active": kind == "session",
        }));
    }
    out
}

fn unavailable(reason: &str) -> Value {
    json!({"available": false, "reason": reason, "staleAfterSeconds": STALE_AFTER})
}

/// The plan's usage windows as Anthropic last reported them.
pub fn collect_limits() -> Value {
    let data = match load() {
        Found::Absent => return unavailable("absent"),
        Found::Unreadable => return unavailable("unreadable"),
        Found::Data(v) => v,
    };

    let cache = match data.get("cachedUsageUtilization") {
        Some(c) if c.is_object() => c,
        _ => return unavailable("absent"),
    };
    let util = match cache.get("utilization") {
        Some(u) if u.is_object() => u,
        _ => return unavailable("absent"),
    };

    let (mut fetched_at, mut age) = (Value::Null, Value::Null);
    if let Some(ms) = cache.get("fetchedAtMs").and_then(|v| v.as_f64()) {
        if let Some(dt) = chrono::DateTime::from_timestamp_millis(ms as i64) {
            fetched_at = json!(crate::collect::iso_of(dt));
        }
        let now = chrono::Utc::now().timestamp() as f64;
        age = json!((now - ms / 1000.0) as i64);
    }

    let account_hash = cache
        .get("accountUuid")
        .and_then(|v| v.as_str())
        .map(|acct| {
            let mut h = Sha256::new();
            h.update(salt().as_bytes());
            h.update(acct.as_bytes());
            format!("{:x}", h.finalize())[..12].to_string()
        })
        .map(Value::from)
        .unwrap_or(Value::Null);

    json!({
        "available": true,
        // Named so the board can say where a number came from without the
        // reader having to take it on trust.
        "source": "~/.claude.json:cachedUsageUtilization",
        "fetchedAt": fetched_at,
        "ageSeconds": age,
        "staleAfterSeconds": STALE_AFTER,
        "accountHash": account_hash,
        "windows": rows(util),
        // Deliberately not computed here: a snapshot can sit in the store for
        // minutes before a browser reads it, and a countdown baked in at
        // collection time would be wrong by exactly that long.
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_scoped_window_uses_only_the_supplied_display_name() {
        let row: Map<String, Value> = serde_json::from_str(
            r#"{"kind":"weekly_scoped","scope":{"model":{"display_name":"Opus"}}}"#,
        )
        .unwrap();
        assert_eq!(label(&row), "Weekly · Opus");
        let bare: Map<String, Value> = serde_json::from_str(r#"{"kind":"weekly_scoped"}"#).unwrap();
        assert_eq!(label(&bare), "Weekly · scoped");
    }

    #[test]
    fn an_unknown_kind_is_titled_not_invented() {
        let row: Map<String, Value> = serde_json::from_str(r#"{"kind":"amber_ladder"}"#).unwrap();
        assert_eq!(label(&row), "Amber Ladder");
    }

    #[test]
    fn the_old_shape_still_reads() {
        let util: Value = serde_json::from_str(
            r#"{"five_hour":{"utilization":42,"resets_at":"2026-08-23T03:00:00Z"}}"#,
        )
        .unwrap();
        let r = rows(&util);
        assert_eq!(r.len(), 1);
        assert_eq!(r[0]["percent"], 42);
        assert_eq!(r[0]["label"], "Session (5h)");
        assert_eq!(r[0]["active"], true);
    }
}
